Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,167 @@
+      the amount of cache flushes that would be driven by releasing a recently feed page, and subsequently allocating
+      another one
+
+
+### **User-space memory allocations**
+
+#### **Facts**
+1. user-space memory must be assumed to not always be backed by physical memory pages
+   meaning that a non running process might have had some of its pages dropped or swapped out to disk
+2. unless some extra effort has been taken, to pin down memory allocations for user-space, it is
+   not really possible to assume much, other than at some point both code that has been executed
+   and data that has been read were present in physical memory. The presence of caching and other
+   optimisations like the zero-page make it very difficult to be assertive beyond this point.
+3. the management of memory pages associated with a process is handled through the process memory
+   map, which consists of several virtual memory areas, which represents address ranges within the
+   process address space, that are put to some use.
+   They come in two flavors:
+   1. anonymous mappings:
+      1. process heap, stack
+      2. zero-initialised variables
+   2. file-backed mappings:
+      1. constants
+      2. the main executable associated with the process
+      3. the runtime linker (optional but typical)
+      4. linked libraries
+4. the kernel uses various optimisations for dealing with processes on-demand mapping:
+   1. a physical page is allocated and mapped to a process only when the process accesses it:
+      1. file backed pages are allocated/mapped only when read/written to
+      2. anonymous pages are allocated only when written to
+   2. zero page: when a page is known to be empty, it is not reserved and mapped; instead,
+      the kernel has 1 specific memory page that is mapped read-only
+   3. shared libraries:
+      when the same library pages are mapped by multiple processes, the library physical pages
+      are mapped as read-only into each process address space
+   4. copy-on-write:
+      1. when a library has own data, this is initially mapped as read-only and shared;
+         only when written to, then a separate physical page is reserved
+      2. same happens for data pages that were treated as zero-page, but then are written to.
+   5. folios: data structures that try to better abstract compound pages and *might* also be used
+      to represent optimised contiguous pages on ARM64 (instead of mapping 16 entries, it is possible
+      to map a 16-pages chunk of physically contiguous memory, that is also aligned to a 16 pages
+      boundary, reducing TLB use).
+      1. A folio acts as intermediary between vma and lower level memory management
+      2. it might pre-allocate/map more pages than explicitly requested
+5. read-ahead: when asked to fetch data from disk, the kernel might attempt to optimise the operation,
+   reading more pages than requested, under the assumption that more requests might be coming soon
```

Comment:
We care because it is even more evidence of the unpredictability of hte kernel with regard to llocation of memory pages.
E.g. if one even attempted to reach a point where transients have completed and all memory had been allocated and put into use, it would not be trivial to verify that such a state has been reached, and to define the steps to reach it, even allowing for some extra ones "just in case".